File: swift/binary_tree.swift
indirect enum Tree<T> {
    case empty
    case node(Tree, T, Tree)

    var size: Int {
        switch self {
            case .empty: return 0
            case .node(let left, _, let right):
                return left.size + 1 + right.size
        }
    }
}

let t: Tree<Int> = .node(.node(.empty, 3, .empty), 7, .empty)
assert(t.size == 2)

File: rust/src/bin/binary_tree.rs
enum Tree<T> {
    Empty,
    Node(Box<Tree<T>>, T, Box<Tree<T>>),
}

use Tree::{Empty, Node};

impl<T> Tree<T> {
    fn size(&self) -> usize {
        match self {
            Empty => 0,
            Node(left, _, right) => left.size() + 1 + right.size(),
        }
    }
}

fn main() {
    let t: Tree<i32> = Node(
        Box::new(Node(Box::new(Empty), 3, Box::new(Empty))),
        7,
        Box::new(Empty),
    );
    assert_eq!(t.size(), 2);
}
